### ISSUE TITLE
Docker Machine settings in `.env` sometimes ignored

### ISSUE THREAD

**Issue by alubbock** [docker/compose#6661]
## Description of the issue
As per [v18.09 docs on .env files](https://docs.docker.com/compose/env-file/), I'm trying to set Docker Machine environment variables automatically by specifying them in a `.env` file (specificially `DOCKER_CERT_PATH`, `DOCKER_HOST`, `DOCKER_TLS_VERIFY`).

Some `docker-compose` commands work (`up`, `down`, `ps`, `logs`), but `docker-compose run` and `docker-compose exec` do not. Example:

```
>>> docker-compose exec nginx pwd
Error: No such container: a05e50bd3c2519b731c860e3c18176acd10feddfb7e48f795c4f065707c54501
```

The above error also occurs with `docker-compose run`. `docker-compose ps -q` confirms that a container with the specified hash is indeed running. Setting the environment variables with `eval $(docker-machine env mymachine)` fixes the issue, so it appears that some `docker-compose` commands are not reading environment variables from the `.env` file.

## Context information (for bug reports)

**Output of `docker-compose version`**
```
docker-compose version 1.23.2, build 1110ad01
docker-py version: 3.6.0
CPython version: 3.6.6
OpenSSL version: OpenSSL 1.1.0h  27 Mar 2018
```

**Output of `docker version`**
```
Client: Docker Engine - Community
 Version:           18.09.2
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        6247962
 Built:             Sun Feb 10 04:12:39 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.2
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.6
  Git commit:       6247962
  Built:            Sun Feb 10 04:13:06 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker-compose config`**
(Make sure to add the relevant `-f` and other flags)
```
services:
  nginx:
    image: nginx:mainline
version: '2.1'
```
**Contents of `.env` file**

```
DOCKER_TLS_VERIFY=1
DOCKER_HOST=tcp://<ip address>:2376
DOCKER_CERT_PATH=/Users/alex/.docker/machine/machines/mymachine
```

## Steps to reproduce the issue

1. Create a Docker Machine (I've only tested AWS - not sure if other platforms are affected):

    docker-machine create --driver amazonec2 --amazonec2-open-port 80 mymachine

2. Create a new directory with the `.env` files and `docker-compose.yml` files as specified above, updating the `.env` file with the `DOCKER_*` environment variables with the values outputted by `docker-machine env mymachine`

3. Run `docker-compose run --rm --entrypoint pwd nginx`

### Observed result

```
Error: No such container: 7527e2716707fd058cf28622ee8229e0ca816454f43098ef49847857c02c0395
```
(the hash changes every time)

### Expected result

```
/
```
(the output of running `pwd` inside the container)

### Stacktrace / full error message

As shown above.

## Additional information

MacOS 10.14.4, running Docker Desktop downloaded from docker.com

**Comment by schmunk42** [comment]
Just stumbled upon the same issue

```
> docker-compose run --rm --entrypoint bash runner-c     
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?


> docker-compose ps
         Name                       Command               State   Ports
-----------------------------------------------------------------------
runner-c_1   /usr/bin/dumb-init /entryp ...   Up   
```

.env file

```
DOCKER_TLS_VERIFY=1
DOCKER_HOST=tcp://myhost.somewhere.de:2376
DOCKER_CERT_PATH=/roj/config/machine/machines/myhost.somewhere.de
DOCKER_MACHINE_NAME=myhost.somewhere.de
```

**Comment by ijc** [comment]
Thanks both for your reports, however I'm unable to reproduce with 1.23.2 nor 1.24.0 on either Linux or MacOS.

Given:
```
$ cat docker-compose.yml 
services:
  nginx:
    image: nginx:mainline
version: '2.1'
$ cat .env 
DOCKER_HOST=tcp://172.16.10.239:2337
```
(where `tcp://172.16.10.239:2337` is deliberate invalid, since this is easier than setting up a cloud node and at least demonstrates that the setting is being found)

I see:
```console
$ ../docker-compose-1.24.0 run --rm --entrypoint pwd nginx
ERROR: Couldn't connect to Docker daemon. You might need to start Docker for Mac.
```
(same with 1.23.2 and  on Linux too, although the error message is different there)

if I remove the `.env` then things work, i.e. it is using the unix socket by default, which I think demonstrates that the `.env` settings are working for me.

Could one if you provide the full output of `docker-compose --verbose run --rm --entrypoint pwd nginx` (i.e. the repro step above, but with `--verbose`) please?

FWIW I get:
```
$ ../docker-compose-1.24.0 --verbose run --rm --entrypoint pwd nginx
compose.config.config.find: Using configuration files: ./docker-compose.yml
docker.utils.config.find_config_file: Trying paths: ['/Users/ijc/.docker/config.json', '/Users/ijc/.dockercfg']
docker.utils.config.find_config_file: Found file at path: /Users/ijc/.docker/config.json
docker.auth.load_config: Couldn't find auth-related section ; attempting to interpret as auth-only file
docker.auth.parse_auth: Config entry for key stackOrchestrator is not auth config
urllib3.connectionpool._new_conn: Starting new HTTP connection (1): 172.16.10.239:2337
Traceback (most recent call last):
  File "site-packages/urllib3/connection.py", line 159, in _new_conn
  File "site-packages/urllib3/util/connection.py", line 80, in create_connection
  File "site-packages/urllib3/util/connection.py", line 70, in create_connection
ConnectionRefusedError: [Errno 61] Connection refused
```
which is expected since that node doesn't listen on that port (but shows the setting is effective)

**Comment by schmunk42** [comment]
> (where tcp://172.16.10.239:2337 is deliberate invalid, since this is easier than setting up a cloud node and at least demonstrates that the setting is being found)

I think it's different since you're trying to connect to a non-existent engine, while we're trying to connect to an existing one.

In my posting above you can see that even for `run` there are requests made to the remote engine, but at some point `docker-compose` tries to use the local socket.

You might be able to reproduce this with a local VM instead of spinning up a cloud node.

Here's the log you've requested: https://gist.github.com/schmunk42/da3085d723215bd4575df6008fad4aff

**Comment by alubbock** [comment]
Here's my log, with the remote IP redacted. I'm getting a different error message to @schmunk42 (`Error: No such container` for me), but I suspect that @schmunk42 either doesn't have a Docker daemon running locally or it requires `sudo` to access, is that right? Looks like the error is occuring at the same `compose.cli.main.call_docker` line for both of us, where a local CLI call is used instead of the HTTP API.

```
docker-compose --verbose run --rm --entrypoint pwd nginx
compose.config.config.find: Using configuration files: ./docker-compose.yml
docker.utils.config.find_config_file: Trying paths: ['/Users/alex/.docker/config.json', '/Users/alex/.dockercfg']
docker.utils.config.find_config_file: Found file at path: /Users/alex/.docker/config.json
docker.auth.load_config: Found 'credsStore' section
urllib3.connectionpool._new_conn: Starting new HTTPS connection (1): <ip address>:2376
urllib3.connectionpool._make_request: https://<ip address>:2376 "GET /v1.24/version HTTP/1.1" 200 558
compose.cli.command.get_client: docker-compose version 1.23.2, build 1110ad01
docker-py version: 3.6.0
CPython version: 3.6.6
OpenSSL version: OpenSSL 1.1.0h  27 Mar 2018
compose.cli.command.get_client: Docker base_url: https://<ip address>:2376
compose.cli.command.get_client: Docker version: Platform={'Name': 'Docker Engine - Community'}, Components=[{'Name': 'Engine', 'Version': '18.09.2', 'Details': {'ApiVersion': '1.39', 'Arch': 'amd64', 'BuildTime': '2019-02-10T03:42:13.000000000+00:00', 'Experimental': 'false', 'GitCommit': '6247962', 'GoVersion': 'go1.10.6', 'KernelVersion': '4.15.0-1032-aws', 'MinAPIVersion': '1.12', 'Os': 'linux'}}], Version=18.09.2, ApiVersion=1.39, MinAPIVersion=1.12, GitCommit=6247962, GoVersion=go1.10.6, Os=linux, Arch=amd64, KernelVersion=4.15.0-1032-aws, BuildTime=2019-02-10T03:42:13.000000000+00:00
compose.cli.verbose_proxy.proxy_callable: docker inspect_network <- ('min_default')
urllib3.connectionpool._make_request: https://<ip address>:2376 "GET /v1.24/networks/min_default HTTP/1.1" 200 545
compose.cli.verbose_proxy.proxy_callable: docker inspect_network -> {'Attachable': True,
 'ConfigFrom': {'Network': ''},
 'ConfigOnly': False,
 'Containers': {},
 'Created': '2019-04-22T20:07:03.955456268Z',
 'Driver': 'bridge',
 'EnableIPv6': False,
 'IPAM': {'Config': [{'Gateway': '172.24.0.1', 'Subnet': '172.24.0.0/16'}],
          'Driver': 'default',
          'Options': None},
...
compose.cli.verbose_proxy.proxy_callable: docker inspect_network <- ('min_default')
urllib3.connectionpool._make_request: https://<ip address>:2376 "GET /v1.24/networks/min_default HTTP/1.1" 200 545
compose.cli.verbose_proxy.proxy_callable: docker inspect_network -> {'Attachable': True,
 'ConfigFrom': {'Network': ''},
 'ConfigOnly': False,
 'Containers': {},
 'Created': '2019-04-22T20:07:03.955456268Z',
 'Driver': 'bridge',
 'EnableIPv6': False,
 'IPAM': {'Config': [{'Gateway': '172.24.0.1', 'Subnet': '172.24.0.0/16'}],
          'Driver': 'default',
          'Options': None},
...
compose.cli.verbose_proxy.proxy_callable: docker inspect_image <- ('nginx:mainline')
urllib3.connectionpool._make_request: https://<ip address>:2376 "GET /v1.24/images/nginx:mainline/json HTTP/1.1" 200 None
compose.cli.verbose_proxy.proxy_callable: docker inspect_image -> {'Architecture': 'amd64',
 'Author': '',
 'Comment': '',
 'Config': {'ArgsEscaped': True,
            'AttachStderr': False,
            'AttachStdin': False,
            'AttachStdout': False,
            'Cmd': ['nginx', '-g', 'daemon off;'],
            'Domainname': '',
            'Entrypoint': None,
...
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=False, filters={'label': ['com.docker.compose.project=min', 'com.docker.compose.service=nginx', 'com.docker.compose.oneoff=False']})
urllib3.connectionpool._make_request: https://<ip address>:2376 "GET /v1.24/containers/json?limit=-1&all=0&size=0&trunc_cmd=0&filters=%7B%22label%22%3A+%5B%22com.docker.compose.project%3Dmin%22%2C+%22com.docker.compose.service%3Dnginx%22%2C+%22com.docker.compose.oneoff%3DFalse%22%5D%7D HTTP/1.1" 200 3
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 0 items)
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=False, filters={'label': ['com.docker.compose.project=min', 'com.docker.compose.service=nginx', 'com.docker.compose.oneoff=False']})
urllib3.connectionpool._make_request: https://<ip address>:2376 "GET /v1.24/containers/json?limit=-1&all=0&size=0&trunc_cmd=0&filters=%7B%22label%22%3A+%5B%22com.docker.compose.project%3Dmin%22%2C+%22com.docker.compose.service%3Dnginx%22%2C+%22com.docker.compose.oneoff%3DFalse%22%5D%7D HTTP/1.1" 200 3
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 0 items)
compose.cli.verbose_proxy.proxy_callable: docker create_host_config <- (links=[], port_bindings={}, binds=[], volumes_from=[], privileged=False, network_mode='min_default', devices=None, dns=None, dns_opt=None, dns_search=None, restart_policy=None, runtime=None, cap_add=None, cap_drop=None, mem_limit=None, mem_reservation=None, memswap_limit=None, ulimits=None, log_config={'Type': '', 'Config': {}}, extra_hosts=None, read_only=None, pid_mode=None, security_opt=None, ipc_mode=None, cgroup_parent=None, cpu_quota=None, shm_size=None, sysctls=None, pids_limit=None, tmpfs=[], oom_kill_disable=None, oom_score_adj=None, mem_swappiness=None, group_add=None, userns_mode=None, init=None, init_path=None, isolation=None, cpu_count=None, cpu_percent=None, nano_cpus=None, volume_driver=None, cpuset_cpus=None, cpu_shares=None, storage_opt=None, blkio_weight=None, blkio_weight_device=None, device_read_bps=None, device_read_iops=None, device_write_bps=None, device_write_iops=None, mounts=None, device_cgroup_rules=None, cpu_period=None, cpu_rt_period=None, cpu_rt_runtime=None)
compose.cli.verbose_proxy.proxy_callable: docker create_host_config -> {'Binds': [],
 'Links': [],
 'LogConfig': {'Config': {}, 'Type': ''},
 'NetworkMode': 'min_default',
 'PortBindings': {},
 'VolumesFrom': []}
compose.cli.verbose_proxy.proxy_callable: docker create_container <- (image='nginx:mainline', volumes={}, command=[], tty=True, stdin_open=True, detach=False, entrypoint='pwd', ports=[], name='min_nginx_run_e12c31a781bf', environment=[], labels={'com.docker.compose.project': 'min', 'com.docker.compose.service': 'nginx', 'com.docker.compose.oneoff': 'True', 'com.docker.compose.slug': 'e12c31a781bfc514df4c28fcad30c6b3c22b0447ff1f15dc52ca454c7990813', 'com.docker.compose.version': '1.23.2'}, host_config={'NetworkMode': 'min_default', 'VolumesFrom': [], 'Binds': [], 'PortBindings': {}, 'Links': [], 'LogConfig': {'Type': '', 'Config': {}}}, networking_config={'EndpointsConfig': {'min_default': {'Aliases': ['nginx'], 'IPAMConfig': {}}}})
urllib3.connectionpool._make_request: https://<ip address>:2376 "POST /v1.24/containers/create?name=min_nginx_run_e12c31a781bf HTTP/1.1" 201 90
compose.cli.verbose_proxy.proxy_callable: docker create_container -> {'Id': 'c39ace7c5d598ea4b73c7eac314374e4bb010eadc756d4b5a5347dbe17596c04',
 'Warnings': None}
compose.cli.verbose_proxy.proxy_callable: docker inspect_container <- ('c39ace7c5d598ea4b73c7eac314374e4bb010eadc756d4b5a5347dbe17596c04')
urllib3.connectionpool._make_request: https://<ip address>:2376 "GET /v1.24/containers/c39ace7c5d598ea4b73c7eac314374e4bb010eadc756d4b5a5347dbe17596c04/json HTTP/1.1" 200 None
compose.cli.verbose_proxy.proxy_callable: docker inspect_container -> {'AppArmorProfile': '',
 'Args': [],
 'Config': {'AttachStderr': True,
            'AttachStdin': True,
            'AttachStdout': True,
            'Cmd': [],
            'Domainname': '',
            'Entrypoint': ['pwd'],
            'Env': ['PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
                    'NGINX_VERSION=1.15.8-1~stretch',
...
compose.cli.verbose_proxy.proxy_callable: docker disconnect_container_from_network <- ('c39ace7c5d598ea4b73c7eac314374e4bb010eadc756d4b5a5347dbe17596c04', 'min_default')
urllib3.connectionpool._make_request: https://<ip address>:2376 "POST /v1.24/networks/min_default/disconnect HTTP/1.1" 200 0
compose.cli.verbose_proxy.proxy_callable: docker disconnect_container_from_network -> None
compose.cli.verbose_proxy.proxy_callable: docker connect_container_to_network <- ('c39ace7c5d598ea4b73c7eac314374e4bb010eadc756d4b5a5347dbe17596c04', 'min_default', aliases=[], ipv4_address=None, ipv6_address=None, links=[], link_local_ips=None)
urllib3.connectionpool._make_request: https://<ip address>:2376 "POST /v1.24/networks/min_default/connect HTTP/1.1" 200 0
compose.cli.verbose_proxy.proxy_callable: docker connect_container_to_network -> None
compose.cli.main.call_docker: /usr/local/bin/docker start --attach --interactive c39ace7c5d598ea4b73c7eac314374e4bb010eadc756d4b5a5347dbe17596c04
Error: No such container: c39ace7c5d598ea4b73c7eac314374e4bb010eadc756d4b5a5347dbe17596c04
compose.cli.verbose_proxy.proxy_callable: docker remove_container <- ('c39ace7c5d598ea4b73c7eac314374e4bb010eadc756d4b5a5347dbe17596c04', force=True, v=True)
urllib3.connectionpool._make_request: https://<ip address>:2376 "DELETE /v1.24/containers/c39ace7c5d598ea4b73c7eac314374e4bb010eadc756d4b5a5347dbe17596c04?v=True&link=False&force=True HTTP/1.1" 204 0
compose.cli.verbose_proxy.proxy_callable: docker remove_container -> None
```

**Comment by ijc** [comment]
> In my posting above you can see that even for run there are requests made to the remote engine, but at some point docker-compose tries to use the local socket.

Thanks for clarifying, that makes sense indeed.

From the logs it looks like maybe the issue correlates with when compose shells out to the `docker` cli instead of calling the API directly. Which leads me to suspect the `call_docker` function in `compose/cli/main.py`. This also correlates with some commands being affected and others not since it is only those which shell out.

Looking at the callers it seems like setting `COMPOSE_INTERACTIVE_NO_CLI=1` might workaround the issue (unless you are on Windows when it doesn't take effect) while we figure out what is going wrong.

**Comment by schmunk42** [comment]
I can confirm that setting `COMPOSE_INTERACTIVE_NO_CLI=1` "fixes" the issue. I am also able to set it in `.env`. Thank you!

**Comment by ijc** [comment]
Should be fixed by #6720 I think.
